### PR TITLE
Vt inbox

### PR DESCRIPTION
**Pull request by stneng**

### CODE REVIEW

File: src/extensions/variable_transfer.rs [hunk 1]
```diff
@@ -17,37 +14,14 @@ impl crate::application::CoLink {
...
         }
-        let params = CreateParams {
-            remote_key_name: format!("_variable_transfer:{}:{}", self.get_task_id()?, key),
-            payload: payload.to_vec(),
-            ..Default::default()
-        };
-        let mut payload = vec![];
-        params.encode(&mut payload).unwrap();
-        self.run_task("remote_storage.create", &payload, &new_participants, false)
+        self._set_variable_remote_storage(key, payload, receivers)
```
Comment (camelop) [inline src/extensions/variable_transfer.rs:24]: we can probably also expose this as an option for external access

File: src/extensions/variable_transfer/p2p_inbox.rs [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+            tls_cert,
+            data_map: data_clone,
+            shutdown_channel: tx,
+            notification_channels: notification_channels_clone,
+        }
+    }
+}
+
+fn gen_cert() -> (Vec<u8>, Vec<u8>) {
+    let subject_alt_names: &[_] = &["vt-p2p.colink".to_string()];
```
[the remark this answers is not in the document]
Comment (camelop) [reply src/extensions/variable_transfer/p2p_inbox.rs:145]: can we separate this to a separate const

[the remark this answers is not in the document]
Comment (camelop) [reply src/extensions/variable_transfer/p2p_inbox.rs:145]: SELF_SIGNED_CERT_DOMAIN_NAME

File: src/extensions/variable_transfer/p2p_inbox.rs [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+
+#[derive(Serialize, Deserialize)]
+pub(crate) struct VTInbox {
+    addr: String,
+    vt_jwt: String,
+    tls_cert: Vec<u8>,
+}
+
+#[derive(Default)]
+pub(crate) struct VTP2P {
```
[the remark this answers is not in the document]
Comment (camelop) [reply src/extensions/variable_transfer/p2p_inbox.rs:160]: VT_P2P_CTX

File: src/extensions/variable_transfer/p2p_inbox.rs [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+
+impl crate::application::CoLink {
+    pub(crate) async fn _set_variable_p2p(
+        &self,
+        key: &str,
+        payload: &[u8],
+        receivers: &[Participant],
+    ) -> Result<(), Error> {
+        for receiver in receivers {
+            // TODO parallel and fallback
```
Comment (camelop) [inline src/extensions/variable_transfer/p2p_inbox.rs:176]: FIX

**Review by camelop — CHANGES_REQUESTED** [review]
Thanks! Please see the detailed comments.

File: src/application.rs [hunk 5]
```diff
@@ -32,7 +32,7 @@ pub struct CoLink {
     pub(crate) ca_certificate: Option<Certificate>,
     pub(crate) identity: Option<Identity>,
     #[cfg(feature = "variable_transfer")]
-    pub(crate) vt_p2p: Arc<crate::extensions::variable_transfer::p2p_inbox::VTP2P>,
+    pub(crate) vt_p2p: Arc<crate::extensions::variable_transfer::p2p_inbox::VTP2PCTX>,
```
Comment (camelop) [inline src/application.rs:35]: Just a thought: should we consider adding underscores? e.g. `VT_P2P_CTX`
Just realized VTP2PCTX looks... hard to parse. What do you think?

File: tests/test_protocol_variable_transfer.rs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            user_id: cls[i].get_user_id()?,
+            role: "receiver".to_string(),
+        });
+    }
+    let data = "test".as_bytes();
+    let task_id = cls[0]
+        .run_task("variable_transfer_test", data, &participants, true)
+        .await?;
+    let idx: usize = rand::thread_rng().gen_range(1..8);
+    let idx2: usize = rand::thread_rng().gen_range(0..8);
```
Comment (camelop) [inline tests/test_protocol_variable_transfer.rs:88]: Are we randomly checking one receiver? Why not check them all?
Also, maybe we can check both w/wo remote storage inside that new for loop?

File: src/protocol.rs [hunk 7]
```diff
@@ -50,6 +50,68 @@ impl CoLinkProtocol {
...
+            }
+            delivery.ack(BasicAckOptions::default()).await.unwrap();
+        }
+
+        Ok(())
+    }
+
+    async fn process_task(&self, task: Task) -> Result<(), Error> {
+        if task.status == "started" {
+            // begin user func
```
Comment (camelop) [inline src/protocol.rs:85]: Is this comment marking still informative?

File: src/extensions/variable_transfer/p2p_inbox.rs [hunk 8]
```diff
@@ -157,10 +142,10 @@ pub(crate) struct VTInbox {
 }
 
 #[derive(Default)]
-pub(crate) struct VTP2P {
+pub(crate) struct VTP2PCTX {
     pub(crate) my_public_addr: Option<String>,
     pub(crate) has_created_inbox: Mutex<bool>,
-    pub(crate) my_inbox: RwLock<Option<MyVTInbox>>,
+    pub(crate) my_inbox: RwLock<Option<VTInboxServer>>,
```
Comment (camelop) [inline src/extensions/variable_transfer/p2p_inbox.rs:148]: do we still need the `my_` prefix? Same for `my_public_addr`

**Review by camelop — CHANGES_REQUESTED** [review]
Nice test! Also, wondering how long does it approximately take to run it? Also, please see the detailed comments.

File: src/application.rs (hunk 5 above)
Comment (stneng) [reply src/application.rs:35]: https://doc.rust-lang.org/1.0.0/style/style/naming/README.html

Comment (camelop) [reply src/application.rs:35]: Yeah, I'm aware of this... so I'm saying VTP2PCTX is hard to parse as a special case.
If you really tend to follow it, maybe we should use `VarTransP2pCtx` or `VtP2pCtx` to follow the CamelCase guideline exactly.

File: tests/test_protocol_variable_transfer.rs [hunk 9]
```diff
@@ -84,19 +83,19 @@ async fn test_vt() -> Result<(), Box<dyn std::error::Error + Send + Sync + 'stat
...
-        .read_or_wait(&format!("tasks:{}:output_remote_storage{}", task_id, idx2))
-        .await?;
-    println!("{}", String::from_utf8_lossy(&res));
-    assert!(res == data);
+    for idx in 1..8 {
+        for idx2 in 0..8 {
+            let res = cls[idx]
+                .read_or_wait(&format!("tasks:{}:output{}", task_id, idx2))
+                .await?;
+            println!("{}", String::from_utf8_lossy(&res));
```
Comment (camelop) [inline tests/test_protocol_variable_transfer.rs:91]: Are we printing too many `"test"`?

**Review by camelop — CHANGES_REQUESTED** [review]
One last naming issue 🍰

File: tests/test_protocol_variable_transfer.rs (hunk 9 above)
Comment (stneng) [reply tests/test_protocol_variable_transfer.rs:91]: `cargo test` will capture the output by default.

Comment (camelop) [reply tests/test_protocol_variable_transfer.rs:91]: Yeah, I know but still, just wondering whether a bunch of "test" would be informative. We can keep it as is for now.

**Review by camelop — APPROVED** [review]
Splendid 🎉
